src/construction.cpp

```diff
@@ -1077,6 +1130,101 @@ void construct::done_window_curtains( const tripoint & )
     g->u.add_msg_if_player( _("After boarding up the window the curtains and curtain rod are left.") );
 }
 
+void construct::done_clean_destroy( const tripoint &p )
+{
+    const auto &ter_bash = g->m.ter( p ).obj().bash;
+    if( ter_bash.str_max < 0 ) {
+        debugmsg( "Tried to destroy an indestructible tile" );
+        return;
+    }
+
+    g->u.add_msg_if_player( _("You safely take down the %s."), g->m.tername( p ).c_str() );
+    g->m.ter_set( p, ter_bash.ter_set );
+    g->m.spawn_items( p, item_group::items_from( ter_bash.drop_group, calendar::turn ) );
+    // @todo Don't force roof ever when in 3D mode
+    g->m.correct_roofs( p, !g->m.has_zlevels() || p.z <= 0 );
+}
+
+void construct::failure_standard( const tripoint & )
+{
+    add_msg( m_info, _( "You cannot build there!" ) );
+}
+
+void construct::failure_no_support( const tripoint &p )
+{
+    const auto &ter = g->m.ter( p ).obj();
+    if( ter.bash.str_max < 0 ||
+        ( !ter.has_flag( TFLAG_SUPPORTS_ROOF ) && !ter.has_flag( TFLAG_COLLAPSES ) ) ) {
+        add_msg( m_info, _( "The %s can't be taken down like that." ), ter.name.c_str() );
+        return;
+    }
+
+    std::map<roof_dir, std::vector<direction>> lacks_support_main;
+    roofs rf( *g, g->m );
+    rf.replace_tiles( {{ p, ter.bash.ter_set }} );
+    for( size_t i = 0; i < (size_t)roof_dir::num_dirs; i++ ) {
+        roof_dir dir_main = static_cast<roof_dir>( i );
+        const tripoint &surr = p + direction_to_offset( dir_main );
+
+        if( g->m.has_flag( TFLAG_COLLAPSES, surr ) && rf.get_support( surr ) <= 0 ) {
+            std::vector<direction> lacks_support_sub;
+            for( size_t j = 0; j < (size_t)roof_dir::num_dirs; j++ ) {
+                roof_dir dir_sec = static_cast<roof_dir>( j );
+                if( rf.get_support( surr, dir_sec ) <= 0 ) {
+                    lacks_support_sub.push_back( convert_directions( dir_sec ) );
+                }
+            }
+            if( lacks_support_sub.size() > 1 ) {
+                lacks_support_main[ dir_main ] = lacks_support_sub;
+            }
+        }
+    }
+
+    add_msg( m_info, _( "You can't take down the %s here because:" ), g->m.tername( p ).c_str() );
+
+    bool support_problem = !lacks_support_main.empty();
+    bool too_weak = ter.bash.str_min > g->u.get_str() + g->u.weapon.damage_melee( DT_BASH );
+
+    if( support_problem ) {
+        add_msg( m_info, _( "Some tiles would end up without support:" ),
+                 ter.name.c_str() );
+        for( const auto &pr : lacks_support_main ) {
+            add_msg( m_info, _( "The %s to the %s: %s" ),
```

Comment:
Better to use positional placeholders here: `The %1$s to the %2$s: %3$s`, because word order of the sentence expressing this is different in some languages, eg. Chinese.